ruby_programming/testing_with_rspec/test_driven_development.md

```diff
@@ -0,0 +1,37 @@
+
+Once the code is written and passes our test suite, you can move onto refactoring. As you might have guessed, running the test suite against the changes we make to our features or units enables us to refactor *while* making sure everything still works.
+
+Colloquially, this process is often referred to as the "**red-green-refactor**" cycle. That's all it is, in a nutshell: automated tests **drive** the design of software. You don't need to know how the entire architecture of your sweet, new, industry-breaking application will work. Your application only has to be broken down, step-by-step, until small units are identified and covered by tests.
+
+## Why Is It Important?
+
+Even if TDD intuitively seems like good practice, its utility is hotly debated even today. 
+
+No one is saying that tests aren't useful. Testing at [all levels](https://robots.thoughtbot.com/rails-test-types-and-the-testing-pyramid) is crucial. Testing ensures that your application is working as intended. Testing provides a safety net that gives your application room to grow and evolve as painlessly as possible as features are added and dependencies change.
+
+The real question is whether or not tests should *drive* development, or whether or not it's important for them to come *first*. Many developers write tests after they've written their code. There is definitely a longer list of pros than cons for having tests in general—no one will disagree with that—so should we test first? 
+
+As usual, it depends. Still, here are some reasons we think it might be important for you to implement TDD:
+
+1. **Practice**. Many employers look for developers who have experience or interest in testing. Sure, you can write tests later, but will you really? Test first for good measure.
+2. **Effectiveness**. Writing tests before code is like stretching before a workout. It's not required, but your workout won't be as good. Writing tests first ensures that the code that comes after works and is of high quality.
+3. **Efficiency**. You'll save time in the long run by not having to re-run entire applications just to make sure little parts work.
+
+## Assignments
+1. Read up on [9 Benefits of Test-Driven Development](https://www.madetech.com/blog/9-benefits-of-test-driven-development) for additional insight on this practice. Note that these points may be more applicable to testing in general rather than TDD specifically, but they are still good points nonetheless.
+2. For a more rounded take, consider DHH's ["TDD is Dead. Long Live TDD"](http://david.heinemeierhansson.com/2014/tdd-is-dead-long-live-testing.html), which sparked controversial conversations on what TDD is, what it has become, and what it should be.
```

Comment:
2. For a more rounded take, consider DHH's ["TDD is Dead. Long Live TDD"](http://david.heinemeierhansson.com/2014/tdd-is-dead-long-live-testing.html), which sparked controversial conversations on what TDD is, what it has become, and what it should be.
</div>